libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewListenerRegistry.kt

```diff
@@ -0,0 +1,84 @@
+package com.mapbox.navigation.dropin
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.dropin.model.ActionInterceptor
+import com.mapbox.navigation.dropin.model.Command
+import com.mapbox.navigation.dropin.model.Store
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+
+@ExperimentalPreviewMapboxNavigationAPI
+internal class NavigationViewListenerRegistry(
+    private val store: Store,
+    private val mapStyleLoader: MapStyleLoader,
+    private val coroutineScope: CoroutineScope
+) {
+    private var listeners = mutableMapOf<NavigationViewListener, Job>()
```

Comment:
Many listeners are held by a `CopyOnWriteArraySet`, that way it can be iterated over and edited without cause IndexOutOfBoundsExceptions. 

But I'm not seeing where this `listeners` is actually iterated over 🤔 . Listeners are only notified upon register?

Reply:
Listeners are not iterated over. Instead, we launch a coroutine for each listener, in which we observe Store state and dispatched actions.

Reply:
Oh interesting, yeah that is creative. Am curious what the performance of this is.. Coroutine context switching is faster than thread context switching. But the Store also queues actions 🤔 . User interactions are meant to be super fast because people can easily detect the time it takes to respond to a button click.

```kotlin
        if (isDispatching) {
            // Actions dispatched during reducer processing loop will be queued up
            // and processed after current action processing finishes.
            pendingActions.add(timedAction)
            return
        }
```

Reply:
Added a little code to measure how long this takes. It varies between 0.1-3.0 milliseconds to dispatch to process. Getting it out to the click listener is harder to measure, but here's a logcat. (measured from Pixel 4XL)

```
2022-06-14 13:01:14.850 23865-23865/com.mapbox.navigation.qa_test_app D/Mapbox: [maps-android\Processed in 0.601146ms : SetDestination(destination=Destination(point=Point{type=Point, bbox=null, coordinates=[-122.28158684557772, 37.804972739766995]}, features=null))]: perf_debug
2022-06-14 13:01:14.851 23865-23865/com.mapbox.navigation.qa_test_app D/Mapbox: [nav-sdk]: [MapboxNavigationViewCustomizedActivity] listener onDestinationChanged = Point{type=Point, bbox=null, coordinates=[-122.28158684557772, 37.804972739766995]}
2022-06-14 13:01:14.851 23865-23865/com.mapbox.navigation.qa_test_app D/Mapbox: [maps-android\Processed in 1.461979ms : SetDestination(destination=Destination(point=Point{type=Point, bbox=null, coordinates=[-122.28158684557772, 37.804972739766995]}, features=null))]: perf_debug
2022-06-14 13:01:14.851 23865-23865/com.mapbox.navigation.qa_test_app D/Mapbox: [maps-android\Processed in 1.579531ms : SetDestination(destination=Destination(point=Point{type=Point, bbox=null, coordinates=[-122.28158684557772, 37.804972739766995]}, features=null))]: perf_debug
2022-06-14 13:01:15.073 23865-23865/com.mapbox.navigation.qa_test_app D/Mapbox: [nav-sdk]: [MapboxNavigationViewCustomizedActivity] listener onDestinationChanged = Point{type=Point, bbox=null, coordinates=[-122.28158684557772, 37.804972739766995]}
```

```
14.851-14.850 = 0.001
15.073-14.851 = 0.222
```

Sometimes it takes 1 millisecond, sometimes it takes 222 milliseconds 😅

Reply:
Interesting. Let's try and optimize it later once we see that this is an actual issue.